File: 백준/Silver/16654. Generalized German Quotation/Generalized German Quotation.py
import sys
from collections import deque
def input(): return sys.stdin.readline().rstrip()

def mkip(ip):
    mk=[]
    if len(ip)%2!=0: return "Keine Loesung"
    for i in range(0,len(ip),2):
        if ip[i]!=ip[i+1]: return "Keine Loesung"
        mk.append(ip[i])
    return ''.join(mk)

def mkans(ip):
    ip=mkip(ip)
    if ip[0]=='K': return ip
    
    dq1=deque()
    dq2=deque()
    op=[]
    
    for c in ip:
        if c=='<':
            if not dq2:
                dq1.append(c)
                op.append('[')
            else:
                dq2.pop()
                op.append(']')
        else:
            if not dq1:
                dq2.append(c)
                op.append('[')
            else:
                dq1.pop()
                op.append(']')
    
    if not dq1 and not dq2: return ''.join(op)
    else: return "Keine Loesung"

ip=input()
print(mkans(ip))

File: 백준/Silver/16654. Generalized German Quotation/Generalized German Quotation.java
import java.io.*;
import java.util.*;

class Main {
    static String noans="Keine Loesung";

    public static String mkip(String ip){
        if(ip.length()%2!=0) return noans;
        
        char[] mk=new char[ip.length()/2];
        for(int i=0,j=0;i<ip.length();i+=2,j++){
            if(ip.charAt(i)!=ip.charAt(i+1)) return noans;
            mk[j]=ip.charAt(i);
        }

        return new String(mk);
    }

    public static String mkans(String tip){
        String ip=mkip(tip);
        if(ip.charAt(0)=='K') return noans;

        Deque<Character> dq1=new ArrayDeque<>(),dq2=new ArrayDeque<>();
        
        StringBuilder ans=new StringBuilder();
        char tc=' ';
        for(char c:ip.toCharArray()){
            if(c=='<'){
                if(dq2.isEmpty()){
                    dq1.offer(c);
                    tc='[';
                }else{
                    dq2.pop();
                    tc=']';
                }
            }else{
                if(dq1.isEmpty()){
                    dq2.offer(c);
                    tc='[';
                }else{
                    dq1.pop();
                    tc=']';
                }
            }
            ans.append(tc);
        }

        if(!dq1.isEmpty() || !dq2.isEmpty()) return noans;
        return ans.toString();
    }

    public static void main(String[] args) throws Exception{
        BufferedReader I=new BufferedReader(new InputStreamReader(System.in));

        System.out.print(mkans(I.readLine()));
    }
}